src/Ops.Service/GroupService.cs

```diff
@@ -0,0 +1,159 @@
+                foreach (var group in allGroups)
+                {
+                    if (!locationGroupIds.Contains(group.Id))
+                    {
+                        groups.Add(group);
+                    }
+                }
+                return groups;
+            }
+            else
+            {
+                throw new OcudaException("There are no Groups created");
+            }
+        }
+
+        public async Task<List<Group>> GetAllGroupsAsync()
+        {
+            return await _groupRepository.GetAllGroupsAsync();
+        }
+
+        public async Task<List<Group>> GetGroupRegions()
+        {
+            return await _groupRepository.GetAllGroupRegions();
+        }
+
+        public async Task<DataWithCount<ICollection<Group>>> GetPaginatedListAsync(
+            BaseFilter filter)
+        {
+            return await _groupRepository.GetPaginatedListAsync(filter);
+        }
+
+        public async Task<Group> GetGroupByIdAsync(int groupId)
+        {
+            return await _groupRepository.FindAsync(groupId);
+        }
+        public async Task<Group> GetGroupByStubAsync(string groupType)
+        {
+            var group = await _groupRepository.GetGroupByStubAsync(groupType);
+            if (group == null)
+            {
+                throw new OcudaException("Group not found.");
+            }
+            else
+            {
+                return group;
+            }
+        }
+        public async Task<Group> AddGroupAsync(Group group)
+        {
+            try
+            {
+                group.GroupType = group.GroupType?.Trim();
+                group.Stub = group.Stub.Trim();
+                group.SubscriptionUrl = group.SubscriptionUrl?.Trim();
+                await ValidateAsync(group);
+                await _groupRepository.AddAsync(group);
+                await _groupRepository.SaveAsync();
+            }
+            catch (OcudaException ex)
+            {
+                _logger.LogError(ex.Message);
+                throw new OcudaException(ex.Message);
+            }
+
+                return group;
+        }
+
+        public async Task<Group> EditAsync(Group group)
+        {
+            try
+            {
+                var currentGroup = await _groupRepository.FindAsync(group.Id);
+                await ValidateAsync(currentGroup);
+                currentGroup.GroupType = group.GroupType?.Trim();
+                currentGroup.Stub = group.Stub.Trim();
+                currentGroup.IsLocationRegion = group.IsLocationRegion;
+                currentGroup.SubscriptionUrl = group.SubscriptionUrl?.Trim();
+                _groupRepository.Update(currentGroup);
+                await _groupRepository.SaveAsync();
+                return currentGroup;
+            }
+            catch (OcudaException ex)
+            {
+                _logger.LogError(ex.Message);
+                throw new OcudaException(ex.Message);
```

Comment:
Why are we creating a new exception here? If we were to rethrow the correct method would be to just put `throw`...but I don't understand why we are catching, logging, and rethrowing.